lib/Tools/circt-bmc/LowerToBMC.cpp

```diff
@@ -0,0 +1,188 @@
+//===- LowerToBMC.cpp -----------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+
+#include "circt/Dialect/Seq/SeqTypes.h"
+#include "circt/Dialect/Verif/VerifOps.h"
+#include "circt/Support/Namespace.h"
+#include "circt/Tools/circt-bmc/Passes.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/LLVMIR/FunctionCallUtils.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/Location.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/APInt.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+
+namespace circt {
+#define GEN_PASS_DEF_LOWERTOBMC
+#include "circt/Tools/circt-bmc/Passes.h.inc"
+} // namespace circt
+
+//===----------------------------------------------------------------------===//
+// Convert Lower To BMC pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct LowerToBMCPass : public circt::impl::LowerToBMCBase<LowerToBMCPass> {
+  using LowerToBMCBase::LowerToBMCBase;
+  void runOnOperation() override;
+};
+} // namespace
+
+static Value lookupOrCreateStringGlobal(OpBuilder &builder, ModuleOp moduleOp,
+                                        StringRef str) {
+  Location loc = moduleOp.getLoc();
+  auto global = moduleOp.lookupSymbol<LLVM::GlobalOp>(str);
+  if (!global) {
+    OpBuilder b = OpBuilder::atBlockEnd(moduleOp.getBody());
+    auto arrayTy = LLVM::LLVMArrayType::get(b.getI8Type(), str.size() + 1);
+    global = b.create<LLVM::GlobalOp>(
+        loc, arrayTy, /*isConstant=*/true, LLVM::linkage::Linkage::Private, str,
+        StringAttr::get(b.getContext(), Twine(str).concat(Twine('\00'))));
+  }
+
+  // FIXME: sanity check the fetched global: do all the attributes match what
+  // we expect?
+
+  return builder.create<LLVM::AddressOfOp>(loc, global);
+}
+
+void LowerToBMCPass::runOnOperation() {
+  Namespace names;
+
+  // Fetch the 'hw.module' operation to model check.
+  Operation *expectedModule = getOperation().lookupSymbol(topModule);
+  if (!expectedModule) {
+    getOperation().emitError("module named '") << topModule << "' not found";
+    return signalPassFailure();
+  }
+  auto hwModule = dyn_cast<hw::HWModuleOp>(expectedModule);
+  if (!hwModule) {
+    expectedModule->emitError("must be a 'hw.module'");
+    return signalPassFailure();
+  }
+
+  if (hwModule.getOps<verif::AssertOp>().empty()) {
+    hwModule.emitError("no property provided to check in module");
```

Comment:
What if an instantiated module as asserts but the top module doesn't? That would still be fine, right?

Reply:
Great point! I guess the low-cost way to handle this would just be to not throw this error if there's an instance op, but I think it's probably nice to catch this early on. If you agree then I'll add a traversal over instances to check whether there's an assertion anywhere?

Reply:
Either way sounds good to me. You'll have to be conservative when hitting an external module or function, though. In that case we could still do a check at runtime, however. Could also be worth skipping for now and adding in a later PR once the general flow is operational 😄

Reply:
Good point - for the time being I'll just leave set it to not throw if there are instance ops